File: benchmarks/crosslang/bench_ptt.py
# Times Python PTT (pypi: parsett) over the golden corpus.
# In-process timing: interpreter startup and corpus loading are excluded,
# so the number is the library's parse cost, comparable to Go ns/op.
#
#   uv venv && uv pip install parsett
#   .venv/bin/python bench_ptt.py ../../parser/testdata/golden.json [passes]

import json
import sys
import time

import PTT


def main() -> None:
    corpus_path = sys.argv[1]
    passes = int(sys.argv[2]) if len(sys.argv) > 2 else 3
    with open(corpus_path, encoding="utf-8") as f:
        titles = [e["title"] for e in json.load(f)]

    for t in titles:  # warmup
        PTT.parse_title(t)

    start = time.perf_counter()
    for _ in range(passes):
        for t in titles:
            PTT.parse_title(t)
    elapsed = time.perf_counter() - start

    per_title_us = elapsed / (passes * len(titles)) * 1e6
    print(
        f"PTT (parsett)\t{len(titles)} titles x {passes}\t{per_title_us:.1f} us/title"
    )


if __name__ == "__main__":
    main()

File: benchmarks/crosslang/bench_ptt.js
// Times the npm parse-torrent-title package over the golden corpus.
// In-process timing: V8 startup and corpus loading are excluded.
//
//   npm install
//   node bench_ptt.js ../../parser/testdata/golden.json [passes]

const fs = require("fs");
const ptt = require("parse-torrent-title");

const corpusPath = process.argv[2];
const passes = parseInt(process.argv[3] || "3", 10);
const titles = JSON.parse(fs.readFileSync(corpusPath, "utf8")).map((e) => e.title);

for (const t of titles) ptt.parse(t); // warmup: let V8 optimize

const start = process.hrtime.bigint();
for (let p = 0; p < passes; p++) {
  for (const t of titles) ptt.parse(t);
}
const elapsedNs = Number(process.hrtime.bigint() - start);

const perTitleUs = elapsedNs / (passes * titles.length) / 1000;
console.log(`parse-torrent-title (npm)\t${titles.length} titles x ${passes}\t${perTitleUs.toFixed(1)} us/title`);
